tests/test_circular_imports.py

```diff
@@ -0,0 +1,110 @@
+"""Tests for circular imports in all local packages and modules.
+
+This ensures all internal packages can be imported right away without
+any need to import some other module before doing so.
+
+This module is based on an idea that pytest uses for self-testing:
+* https://github.com/aio-libs/aiohttp/blob/91108c9/tests/test_circular_imports.py
+* https://github.com/sanitizers/octomachinery/blob/be18b54/tests/circular_imports_test.py
+* https://github.com/pytest-dev/pytest/blob/d18c75b/testing/test_meta.py
+* https://twitter.com/codewithanthony/status/1229445110510735361
+"""
+
+from __future__ import annotations
+
+import os
+import pkgutil
+import subprocess
+import sys
+from collections.abc import Iterator
+from itertools import chain
+from pathlib import Path
+from types import ModuleType
+
+import pytest
+
+import piptools
+from piptools.utils import PIP_VERSION
+
+
+def _find_all_importables(pkg: ModuleType) -> list[str]:
+    """Find all importables in the project.
+
+    Return them in order.
+    """
+    return sorted(
+        set(
+            chain.from_iterable(
+                _discover_path_importables(Path(p), pkg.__name__) for p in pkg.__path__
+            ),
+        ),
+    )
+
+
+def _discover_path_importables(pkg_pth: Path, pkg_name: str) -> Iterator[str]:
+    """Yield all importables under a given path and package."""
+    for dir_path, _d, file_names in os.walk(pkg_pth):
+        pkg_dir_path = Path(dir_path)
+
+        if pkg_dir_path.parts[-1] == "__pycache__":
+            continue
+
+        if all(Path(_).suffix != ".py" for _ in file_names):
```

Comment:
This is normally not hit, IIRC.
```suggestion
        if all(Path(_).suffix != ".py" for _ in file_names):  # pragma: no cover
```